### PR TITLE
core: hle: kernel: Updated implementation of svcSetHeapSize.

### PR DESCRIPTION
**Pull request by bunnei**
- Updates our svcSetHeapSize with latest HOS, furthermore allowing heap size to properly be extended/shrunk.
- Validated with tests https://github.com/Atmosphere-NX/Atmosphere/blob/master/tests/TestSvc/source/test_set_heap_size.cpp.

### CODE REVIEW

File: src/core/hle/kernel/k_page_table.cpp [hunk 1]
```diff
@@ -859,61 +859,125 @@ ResultCode KPageTable::SetMemoryAttribute(VAddr addr, std::size_t size, KMemoryA
...
+            block_manager->Update(heap_region_start + size, num_pages, KMemoryState::Free,
+                                  KMemoryPermission::None, KMemoryAttribute::None);
+
+            // Update the current heap end.
+            current_heap_end = heap_region_start + size;
+
+            // Set the output.
+            *out = heap_region_start;
+            return ResultSuccess;
+        } else if (size == static_cast<std::size_t>(current_heap_end - heap_region_start)) {
```
Comment (lioncash) [inline src/core/hle/kernel/k_page_table.cpp:919]:
```suggestion
        } else if (size == GetHeapSize()) {
```

File: src/core/hle/kernel/k_page_table.cpp [hunk 2]
```diff
@@ -859,61 +859,125 @@ ResultCode KPageTable::SetMemoryAttribute(VAddr addr, std::size_t size, KMemoryA
...
+        // Lock the table.
+        std::lock_guard lk(page_table_lock);
+
+        // Validate that setting heap size is possible at all.
+        R_UNLESS(!is_kernel, ResultOutOfMemory);
+        R_UNLESS(size <= static_cast<std::size_t>(heap_region_end - heap_region_start),
+                 ResultOutOfMemory);
+        R_UNLESS(size <= max_heap_size, ResultOutOfMemory);
+
+        if (size < static_cast<std::size_t>(current_heap_end - heap_region_start)) {
```
Comment (lioncash) [inline src/core/hle/kernel/k_page_table.cpp:888]:
```suggestion
        if (size < GetHeapSize()) {
```

File: src/core/hle/kernel/k_page_table.cpp [hunk 3]
```diff
@@ -859,61 +859,125 @@ ResultCode KPageTable::SetMemoryAttribute(VAddr addr, std::size_t size, KMemoryA
...
+            return ResultSuccess;
+        } else if (size == static_cast<std::size_t>(current_heap_end - heap_region_start)) {
+            // The size requested is exactly the current size.
+            *out = heap_region_start;
+            return ResultSuccess;
+        } else {
+            // We have to allocate memory. Determine how much to allocate and where while the table
+            // is locked.
+            cur_address = current_heap_end;
+            allocation_size = size - (current_heap_end - heap_region_start);
```
Comment (lioncash) [inline src/core/hle/kernel/k_page_table.cpp:927]:
```suggestion
            allocation_size = size - GetHeapSize();
```

File: src/core/hle/kernel/k_page_table.cpp [hunk 4]
```diff
@@ -859,61 +859,125 @@ ResultCode KPageTable::SetMemoryAttribute(VAddr addr, std::size_t size, KMemoryA
...
+            // Validate memory state.
+            std::size_t num_allocator_blocks;
+            R_TRY(this->CheckMemoryState(
+                std::addressof(num_allocator_blocks), heap_region_start + size,
+                (current_heap_end - heap_region_start) - size, KMemoryState::All,
+                KMemoryState::Normal, KMemoryPermission::All, KMemoryPermission::ReadAndWrite,
+                KMemoryAttribute::All, KMemoryAttribute::None));
+
+            // Unmap the end of the heap.
+            const auto num_pages = ((current_heap_end - heap_region_start) - size) / PageSize;
```
Comment (lioncash) [inline src/core/hle/kernel/k_page_table.cpp:901]:
```suggestion
            const auto num_pages = (GetHeapSize() - size) / PageSize;
```

File: src/core/hle/kernel/k_page_table.cpp [hunk 5]
```diff
@@ -859,61 +859,125 @@ ResultCode KPageTable::SetMemoryAttribute(VAddr addr, std::size_t size, KMemoryA
...
+
+        if (size < static_cast<std::size_t>(current_heap_end - heap_region_start)) {
+            // The size being requested is less than the current size, so we need to free the end of
+            // the heap.
+
+            // Validate memory state.
+            std::size_t num_allocator_blocks;
+            R_TRY(this->CheckMemoryState(
+                std::addressof(num_allocator_blocks), heap_region_start + size,
+                (current_heap_end - heap_region_start) - size, KMemoryState::All,
```
Comment (lioncash) [inline src/core/hle/kernel/k_page_table.cpp:896]:
```suggestion
                GetHeapSize() - size, KMemoryState::All,
```

File: src/core/hle/kernel/k_page_table.h [hunk 6]
```diff
@@ -183,14 +183,15 @@ class KPageTable final : NonCopyable {
     constexpr VAddr GetAliasCodeRegionSize() const {
         return alias_code_region_end - alias_code_region_start;
     }
+    size_t GetNormalMemorySize() {
+        std::lock_guard lk(page_table_lock);
+        return (current_heap_end - heap_region_start) + mapped_physical_memory_size;
```
Comment (lioncash) [inline src/core/hle/kernel/k_page_table.h:188]:
```suggestion
        return GetHeapSize() + mapped_physical_memory_size;
```

File: src/core/hle/kernel/k_page_table.h [hunk 7]
```diff
@@ -183,14 +183,15 @@ class KPageTable final : NonCopyable {
     constexpr VAddr GetAliasCodeRegionSize() const {
         return alias_code_region_end - alias_code_region_start;
     }
+    size_t GetNormalMemorySize() {
```
Comment (lioncash) [inline src/core/hle/kernel/k_page_table.h:186]:
```suggestion
    size_t GetNormalMemorySize() const {
```

File: src/core/hle/kernel/k_page_table.h [hunk 8]
```diff
@@ -183,14 +183,15 @@ class KPageTable final : NonCopyable {
...
+        std::lock_guard lk(page_table_lock);
+        return (current_heap_end - heap_region_start) + mapped_physical_memory_size;
+    }
     constexpr std::size_t GetAddressSpaceWidth() const {
         return address_space_width;
     }
     constexpr std::size_t GetHeapSize() {
```
Comment (lioncash) [inline src/core/hle/kernel/k_page_table.h:193]:
```suggestion
    constexpr std::size_t GetHeapSize() const {
```

File: src/core/hle/kernel/k_page_table.h (hunk 7 above)
Comment (bunnei) [reply src/core/hle/kernel/k_page_table.h:186]: Cannot be const, as it takes the lock below

**Pull request merged**